src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/SampleNameMap.java

```diff
@@ -0,0 +1,274 @@
+                    throw new UserException("Malformed URI: " + e.toString());
+                }
+            }
+        } catch (final IOException e) {
+            throw new UserException.CouldNotReadInputFile(sampleToFileMapPath, "exception while reading sample->filename mapping file",  e);
+        }
+    }
+
+    /**
+     * Tests whether the sample name is legal. Sample names must be non-empty, and
+     * may have internal whitespace but not leading/trailing whitespace.
+     *
+     * @param sampleName sample name to test
+     * @return true if sampleName is legal, otherwise false
+     */
+    private boolean sampleNameIsLegal(final String sampleName) {
+        return sampleName != null &&
+                ! sampleName.trim().isEmpty() &&
+                sampleName.trim().equals(sampleName);
+    }
+
+    /**
+     * Add a new sample mapping
+     *
+     * @param sampleName name of the sample
+     * @param vcfPath path to the VCF for the sample
+     */
+    public void addSample(final String sampleName, final URI vcfPath) {
+        addSample(sampleName, vcfPath, null);
+    }
+
+    /**
+     * Add a new sample mapping
+     *
+     * @param sampleName name of the sample
+     * @param vcfPath path to the VCF for the sample (not null)
+     * @param vcfIndexPath path to the index for the sample (may be null)
+     */
+    public void addSample(final String sampleName, final URI vcfPath, final URI vcfIndexPath) {
+        if ( ! sampleNameIsLegal(sampleName) ) {
+            throw new UserException.BadInput("Sample name " + sampleName + " is not legal. Sample names must be non-empty and not contain leading or trailing whitespace");
+        }
+        if ( vcfPath == null ) {
+            throw new UserException.BadInput("VCF path for sample " + sampleName + " was null");
+        }
+
+        final URI previousPath = sampleNameToVcfPath.put(sampleName, vcfPath);
+        if (previousPath != null) {
+            throw new UserException.BadInput("Duplicate sample: " + sampleName + ". Sample was found in both "
+                    + vcfPath + " and " + previousPath + ".");
+        }
+
+        if (vcfIndexPath != null) {
+            final URI previousIndexPath = sampleNameToVcfIndexPath.put(sampleName, vcfIndexPath);
+            if (previousIndexPath != null) {
+                throw new UserException.BadInput("For sample " + sampleName + ", attempted to specify multiple indices: " + vcfIndexPath + " and " + previousIndexPath);
+            }
+        }
+    }
+
+    /**
+     * @return The full mapping of sample names -> VCF paths, with the sample names in sorted order
+     */
+    public SortedMap<String, URI> getSampleNameToVcfPath() {
+        return sampleNameToVcfPath;
+    }
+
+    /**
+     * @param sample sample name
+     * @return the VCF associated with that sample name, as a URI
+     */
+    public URI getVCFForSample(final String sample) {
+        return sampleNameToVcfPath.get(sample);
```

Comment:
Do we really need both this and the one below it?  It's only called in tests.

Reply:
Should we validate that they can be loaded paths on load instead of failing partway through?  Or do we want to allow invalid paths to support genomicsdb's use of providers we don't support.  Should we be passing through without going to path first?